[AsyncTextInput] Pressing space twice after accepting an inline prediction no longer inserts a full stop

whsieh opened the pull request:
#### 2a6ce5d4b0c461264eab28f28c7856460a83e465
<pre>
[AsyncTextInput] Pressing space twice after accepting an inline prediction no longer inserts a full stop
<a href="https://bugs.webkit.org/show_bug.cgi?id=267653">https://bugs.webkit.org/show_bug.cgi?id=267653</a>
<a href="https://rdar.apple.com/120847806">rdar://120847806</a>

Reviewed by Megan Gardner.

When `UIAsyncTextInput` is enabled, most calls to `-deleteBackward` and `-_deleteForwardAndNotify:`
have been merged into `-deleteInDirection:toGranularity:`, with `UITextGranularityCharacter` as the
given granularity. However, there was a subtle behavior change here: we now notify the system input
delegate (i.e. `UIKeyboardImpl`) by calling into `-selectionWillChange:` / `-selectionDidChange:`
when deleting backwards by a single character, whereas we didn&apos;t before.

Notifying the system delegate causes the keyboard to clear out some cached keyboard input state,
including a member that&apos;s used to determine whether or not the last inserted character was a space:
`m_previousInputString`. This previous input string is consulted by UIKit when deciding whether to
insert a space or to replace a recently-inserted space with a period character; because of the new
delegate notifications, we now always choose the former instead of the latter when performing a
double-space.

To limit risk, we address this regression by restoring shipping behavior, and avoid notifying the
keyboard about selection changes when performing the equivalent of `-deleteBackward`.

Unfortunately, I couldn&apos;t figure out a reliable way to write a test for this bug. For now, I filed
&lt;<a href="https://webkit.org/b/267654">https://webkit.org/b/267654</a>&gt; for this.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _executeEditCommand:]):
(-[WKContentView _executeEditCommand:notifyDelegate:]):
(-[WKContentView deleteInDirection:toGranularity:]):

Canonical link: <a href="https://commits.webkit.org/273158@main">https://commits.webkit.org/273158@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77b00d0c2459a53c4b1171a8c4a3c1debf45eb94

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34475 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13294 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36477 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37161 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31174 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15684 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10401 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30163 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34996 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11245 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30736 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9830 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9910 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38439 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31308 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31111 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35979 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10033 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7919 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33927 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11851 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7928 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/10589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10894 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -7427,15 +7429,21 @@ - (BOOL)isKeyboardScrollingAnimationRunning
-        [view _internalEndSelectionChange];
+    if (notifyDelegate == WebKit::NotifyInputDelegate::Yes)
+        [self _internalBeginSelectionChange];
+    _page->executeEditCommand(commandName, { }, [view = retainPtr(self), notifyDelegate] {
```
just curious; What's the change in this `view` capture about?

whsieh replied:
Ah, just decided to make this a bit more concise while I was here.

aprotyas reviewed (commented):
LGTM

whsieh committed:
[AsyncTextInput] Pressing space twice after accepting an inline prediction no longer inserts a full stop
https://bugs.webkit.org/show_bug.cgi?id=267653
rdar://120847806

Reviewed by Megan Gardner.

When `UIAsyncTextInput` is enabled, most calls to `-deleteBackward` and `-_deleteForwardAndNotify:`
have been merged into `-deleteInDirection:toGranularity:`, with `UITextGranularityCharacter` as the
given granularity. However, there was a subtle behavior change here: we now notify the system input
delegate (i.e. `UIKeyboardImpl`) by calling into `-selectionWillChange:` / `-selectionDidChange:`
when deleting backwards by a single character, whereas we didn't before.

Notifying the system delegate causes the keyboard to clear out some cached keyboard input state,
including a member that's used to determine whether or not the last inserted character was a space:
`m_previousInputString`. This previous input string is consulted by UIKit when deciding whether to
insert a space or to replace a recently-inserted space with a period character; because of the new
delegate notifications, we now always choose the former instead of the latter when performing a
double-space.

To limit risk, we address this regression by restoring shipping behavior, and avoid notifying the
keyboard about selection changes when performing the equivalent of `-deleteBackward`.

Unfortunately, I couldn't figure out a reliable way to write a test for this bug. For now, I filed
<https://webkit.org/b/267654> for this.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _executeEditCommand:]):
(-[WKContentView _executeEditCommand:notifyDelegate:]):
(-[WKContentView deleteInDirection:toGranularity:]):

Canonical link: https://commits.webkit.org/273158@main
